### PR TITLE
Share samples across cases

### LINKED ISSUE #122: Modify the database so one sample can belong to several cases
This affects when you delete cases

### PR DESCRIPTION
**Pull request by northwestwitch**
### This PR adds | fixes:
- Modify database so that samples can belong to various cases (fix #122 )

### How to test:
- Using the test database or a local database, create a case that is connected to 2 samples
- Delete one fo these samples and make sure that case is till there
- Delete the other sample and verify that now the sample is gone

### Expected outcome:
- [x] The functionality should be working as described

### Review:
- [ ] Code approved by
- [x] Tests executed by CR

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner

### CODE REVIEW

**Comment by northwestwitch** [comment]
- Given a local database, when launching the app then the many-to-many linking table case_sample is created:

image


- Insert 2 cases: case1 and case2 using the endpoint:

image

- Create one sample connected to the 2 cases and make sure it is returned when cases are queried:

image


```
[
  {
    "display_name": "Case 1",
    "name": "case1",
    "id": 1,
    "samples": [
      {
        "id": 1,
        "display_name": "Shared Case",
        "coverage_file_path": "/Users/chiararasi/Documents/work/d4_data/hg002.d4",
        "created_at": "2023-05-25T10:03:17",
        "track_name": "shared_case",
        "name": "shared_case"
      }
    ]
  },
  {
    "display_name": "Case 2",
    "name": "case2",
    "id": 2,
    "samples": [
      {
        "id": 1,
        "display_name": "Shared Case",
        "coverage_file_path": "/Users/chiararasi/Documents/work/d4_data/hg002.d4",
        "created_at": "2023-05-25T10:03:17",
        "track_name": "shared_case",
        "name": "shared_case"
      }
    ]
  }
]
```

- When I remove the first case:

image

- Then the sample is still there:

image

- Associated to Case 2:

image

- When I remove also Case 2:

image

- Then also the sample is gone:

image

**Comment by northwestwitch** [comment]
Took me a while since this database change had many repercussions across the code, but it's ready for review now!

File: src/chanjo2/models/sql_models.py [hunk 1]
```diff
@@ -1,37 +1,39 @@
-from sqlalchemy import Column, DateTime, Enum, ForeignKey, Integer, String, Index
+from sqlalchemy import Column, DateTime, Enum, ForeignKey, Integer, String, Index, Table
 from sqlalchemy.orm import relationship
 from sqlalchemy.sql import func
 
 from chanjo2.dbutil import Base
 from chanjo2.models.pydantic_models import Builds
 
+CaseSample = Table(
```
Comment (northwestwitch) [inline src/chanjo2/models/sql_models.py:8]: This is a new table that connects cases and sample in a many-to-many relationship. Followed this tutorial: https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#many-to-many

File: src/chanjo2/crud/cases.py [hunk 2]
```diff
@@ -38,9 +41,38 @@ def count_cases(db: Session) -> int:
     return db.query(SQLCase).count()
 
 
+def delete_entry(db: Session, table: Union[SQLCase, SQLSample], id: int) -> int:
+    """Deletes a Sample or Case entry from the database"""
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/cases.py:45]:
```suggestion
    """Deletes a Sample or Case entry from the database."""
```

File: src/chanjo2/crud/samples.py [hunk 3]
```diff
@@ -40,10 +34,10 @@ def get_samples_by_name(db: Session, sample_names: List[str]) -> List[SQLSample]
 def get_case_samples(db: Session, case_name: str) -> List[SQLSample]:
```
Comment (henrikstranneheim) [inline src/chanjo2/crud/samples.py:34]: Return hint should use Optional now

File: src/chanjo2/endpoints/samples.py [hunk 4]
```diff
@@ -22,13 +22,12 @@ def create_sample_for_case(
     db: Session = Depends(get_session),
 ):
     """Add a sample to a case in the database."""
```
Comment (henrikstranneheim) [inline src/chanjo2/endpoints/samples.py:24]: Add return hint

File: src/chanjo2/endpoints/samples.py [hunk 5]
```diff
@@ -22,13 +22,12 @@ def create_sample_for_case(
...
-    db_sample = create_sample_in_case(db=db, sample=sample)
-    if db_sample is None:
+    result: Union[SQLSample, str] = create_sample_in_case(db=db, sample=sample)
```
Comment (henrikstranneheim) [inline src/chanjo2/endpoints/samples.py:25]: I think db_sample was clearer and you are raising if it is not an object so it seems safe

**Review by henrikstranneheim — APPROVED** [review]
Awesome 🤩! I think doing this now saved us a lot of future headache 👍

File: src/chanjo2/endpoints/samples.py (hunk 5 above)
Comment (northwestwitch) [reply src/chanjo2/endpoints/samples.py:25]: > Awesome 🤩! I think doing this now saved us a lot of future headache 👍

I agree!

File: src/chanjo2/endpoints/samples.py (hunk 4 above)
Comment (northwestwitch) [reply src/chanjo2/endpoints/samples.py:24]: There is already this line that tells you what is returned:

`@router.post("/samples/", response_model=Sample)`

Was it this what you meant?

File: src/chanjo2/endpoints/samples.py (hunk 5 above)
Comment (northwestwitch) [reply src/chanjo2/endpoints/samples.py:25]: Yes, but this way I am able to provide a precise error message: return can be either a sql sample or an error message like this: 
- return f"Could not find a case with name: {sample.case_name}"
- "Sample already connected to given case."

I think knowing why creating the sample failed is very important

File: src/chanjo2/crud/samples.py (hunk 3 above)
Comment (northwestwitch) [reply src/chanjo2/crud/samples.py:34]: A list is returned either way, it's either List[SQLSample] or List (empty list).

I will simplify and put only list in return type

**Comment by northwestwitch** [comment]
I think it would be best to expand the tests for this functionality, adding automatic tests that cover the situations I tested manually in my previous comment. But I'll do it in another PR!

**Pull request merged**